token-tracker/src/truth-sources/lp-tokens.js

```diff
@@ -0,0 +1,275 @@
+
+// https://info.uniswap.org/pair/0xe6f19dab7d43317344282f803f8e8d240708174a
+const KEEPETH_PAIR_ADDRESS = "0xe6f19dab7d43317344282f803f8e8d240708174a"
+// https://info.uniswap.org/pair/0x38c8ffee49f286f25d25bad919ff7552e5daf081
+const KEEPTBTC_PAIR_ADDRESS = "0x38c8ffee49f286f25d25bad919ff7552e5daf081"
+
+// https://etherscan.io/tx/0xc64ac175846e719bb4f7f9b17a0b04bc365db3dda9d97ef70d7ede8f9c1a265b
+const KEEPETH_CREATION_BLOCK = "10100034"
+// https://etherscan.io/tx/0x1592f9b235c602c87a5b8cc5f896164dc43d16b92664cb9c8b420d28b64ca4a0
+const KEEPTBTC_CREATION_BLOCK = "11452642"
+
+const KEEP_IN_LP_KEEPETH_BALANCES_PATH =
+  "./tmp/keep-in-lp-keepeth-token-balances.json"
+
+const KEEP_IN_LP_KEEPTBTC_BALANCES_PATH =
+  "./tmp/keep-in-lp-keeptbtc-token-balances.json"
+
+export class LPTokenTruthSource extends ITruthSource {
+  /**
+   * @param {Context} context
+   * @param {Number} targetBlock
+   */
+  constructor(context, targetBlock) {
+    super(context, targetBlock)
+  }
+
+  async initialize() {
+    const chainID = await getChainID(this.context.web3)
+
+    const keepEthTokenContract = EthereumHelpers.buildContract(
+      this.context.web3,
+      JSON.parse(keepEthTokenJson.result),
+      KEEPETH_PAIR_ADDRESS
+    )
+
+    const lpRewardKeepEthContract = EthereumHelpers.buildContract(
+      this.context.web3,
+      LPRewardsKEEPETHJson.abi,
+      LPRewardsKEEPETHJson.networks[chainID].address
+    )
+
+    const keepTbtcTokenContract = EthereumHelpers.buildContract(
+      this.context.web3,
+      JSON.parse(keepTbtcTokenJson.result),
+      KEEPTBTC_PAIR_ADDRESS
+    )
+
+    const lpRewardKeepTbtcContract = EthereumHelpers.buildContract(
+      this.context.web3,
+      LPRewardsKEEPTBTCJson.abi,
+      LPRewardsKEEPTBTCJson.networks[chainID].address
+    )
+
+    this.liquidityStakingObjects = {
+      KEEPETH: {
+        lpTokenContract: keepEthTokenContract,
+        lpRewardsContract: lpRewardKeepEthContract,
+        lpCreationBlock: KEEPETH_CREATION_BLOCK,
+        keepInLpTokenFilePath: KEEP_IN_LP_KEEPETH_BALANCES_PATH,
+        lpPairAddress: KEEPETH_PAIR_ADDRESS,
+      },
+      KEEPTBTC: {
+        lpTokenContract: keepTbtcTokenContract,
+        lpRewardsContract: lpRewardKeepTbtcContract,
+        lpCreationBlock: KEEPTBTC_CREATION_BLOCK,
+        keepInLpTokenFilePath: KEEP_IN_LP_KEEPTBTC_BALANCES_PATH,
+        lpPairAddress: KEEPTBTC_PAIR_ADDRESS,
+      },
+    }
+  }
+
+  /**
+   * Finds all historic stakers of LP KEEP-ETH / KEEP-TBTC pair token based on
+   * "Transfer" events
+   *
+   * @param {String} pairName LP pair name
+   * @param {Object} pairObj LP pair object
+   *
+   * @return {Set<Address>} All historic LP KEEP-ETH / KEEP-TBTC token stakers
+   * */
+  async findStakers(pairName, pairObj) {
+    const lpRewardsContractAddress = pairObj.lpRewardsContract.options.address
+
+    logger.info(
+      `looking for Transfer events emitted from ${lpRewardsContractAddress} ` +
```

Comment:
Is this info log accurate? 

We are looking for Transfer events emitted from the LP token contract (not LP rewards contract) and then we filter by to set to LP rewards contract.

I'd rather log here that we are looking for all LP token Transfer events between blocks X and Y and then, log how many of them were filtered out at line 121.

Reply:
Actually, I was thinking about refactoring this a bit in the upcoming commits. I kind of fixed my mind around LP Token transfers, but what I can check instead are "Staked" events on LPRewards contracts. It will narrow the events window even further and also get rid of filter that checks if these LP tokens were transferred to LP rewards.

Reply:
230ffa5